core/solver/pipe_cg.cpp

```diff
@@ -0,0 +1,301 @@
+{
+    return build()
+        .with_generated_preconditioner(
+            share(as<Transposable>(this->get_preconditioner())->transpose()))
+        .with_criteria(this->get_stop_criterion_factory())
+        .on(this->get_executor())
+        ->generate(
+            share(as<Transposable>(this->get_system_matrix())->transpose()));
+}
+
+
+template <typename ValueType>
+std::unique_ptr<LinOp> PipeCg<ValueType>::conj_transpose() const
+{
+    return build()
+        .with_generated_preconditioner(share(
+            as<Transposable>(this->get_preconditioner())->conj_transpose()))
+        .with_criteria(this->get_stop_criterion_factory())
+        .on(this->get_executor())
+        ->generate(share(
+            as<Transposable>(this->get_system_matrix())->conj_transpose()));
+}
+
+
+template <typename ValueType>
+void PipeCg<ValueType>::apply_impl(const LinOp* b, LinOp* x) const
+{
+    if (!this->get_system_matrix()) {
+        return;
+    }
+    experimental::precision_dispatch_real_complex_distributed<ValueType>(
+        [this](auto dense_b, auto dense_x) {
+            this->apply_dense_impl(dense_b, dense_x);
+        },
+        b, x);
+}
+
+
+template <typename ValueType>
+template <typename VectorType>
+void PipeCg<ValueType>::apply_dense_impl(const VectorType* dense_b,
+                                         VectorType* dense_x) const
+{
+    using std::swap;
+    using LocalVector = matrix::Dense<ValueType>;
+
+    constexpr uint8 RelativeStoppingId{1};
+
+    auto exec = this->get_executor();
+    this->setup_workspace();
+
+    GKO_SOLVER_VECTOR(r, dense_b);
+    GKO_SOLVER_VECTOR(z, dense_b);
+    GKO_SOLVER_VECTOR(p, dense_b);
+    GKO_SOLVER_VECTOR(w, dense_b);
+    GKO_SOLVER_VECTOR(m, dense_b);
+    GKO_SOLVER_VECTOR(n, dense_b);
+    GKO_SOLVER_VECTOR(q, dense_b);
+    GKO_SOLVER_VECTOR(f, dense_b);
+    GKO_SOLVER_VECTOR(g, dense_b);
+
+    GKO_SOLVER_SCALAR(beta, dense_b);
+    GKO_SOLVER_SCALAR(delta, dense_b);
+    GKO_SOLVER_SCALAR(prev_rho, dense_b);
+    GKO_SOLVER_SCALAR(rho, dense_b);
+
+    GKO_SOLVER_ONE_MINUS_ONE();
+
+    bool one_changed{};
+
+    GKO_SOLVER_STOP_REDUCTION_ARRAYS();
+
+    // r = b
+    // prev_rho = 1.0
+    exec->run(pipe_cg::make_initialize_1(gko::detail::get_local(dense_b),
+                                         gko::detail::get_local(r), prev_rho,
+                                         &stop_status));
+    // r = r - Ax
+    this->get_system_matrix()->apply(neg_one_op, dense_x, one_op, r);
+    // z = preconditioner * r
+    this->get_preconditioner()->apply(r, z);
+    // w = A * z
+    this->get_system_matrix()->apply(z, w);
+    // m = preconditioner * w
+    this->get_preconditioner()->apply(w, m);
+    // n = A * m
+    this->get_system_matrix()->apply(m, n);
+    // rho = dot(r, z)
+    r->compute_conj_dot(z, rho, reduction_tmp);
+    // delta = dot(w, z)
+    w->compute_conj_dot(z, delta, reduction_tmp);
+    // can these two dot products above be merged in some way?
```

Comment:
```suggestion
    // TODO: can these two dot products above be merged in some way?
```
the current support should allow [r w] and [z z], so you need to duplicate the z unfortunately.
It might be possible if we introduce conjugate transposed apply on dense it will be like [r w]' * z

Reply:
Yes, that would be the plan, but we will do that in a later PR.

Reply:
@pratikvn so should I remove this comment for now? I've put it there so that we don't forget about it

Reply:
I would say we can keep that for reminder